{
 "metadata": {
  "name": "",
  "signature": "sha256:cc3b8832969f972277ca518abefc6340a2a5b6646ee94a8c876fcacf2f3aaa79"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Introduction\n",
      "\n",
      "Pathogen.jl is a package that provides utilities for the simulation and inference of pathogen phylodynamics, built in the [Julia language](http://julialang.org). Specifically, Pathogen.jl presents an extension to the individual level infectious diesase transmission models (ILMs) of Deardon et al. (2010), to simultaneously model infectious disease transmission and evolution. Pathogen genomic sequences are used in conjunction with the covariate and disease state information of individuals to infer disease transmission pathways, external disease pressure, and infectivity parameters.\n",
      "\n",
      "At it's core, an ILM describes the probability an individual will become exposed to a disease in a specific time interval based on their covariate information and the disease state of the rest of the population. In a discrete time ILM, the probability an individual $i$ will become exposed at time $t$ is as (Deardon et al. 2010):\n",
      "\n",
      "$P(i,t) = \\begin{cases} 1 - \\exp \\left \\{ \\left[-\\Omega_{S}(i)\\sum_{j \\in I_{(t)}} \\Omega_{T}(j)\\kappa(i,j)\\right] + \\epsilon(i,t) \\right \\} \\text{ if } i \\in S_{(t)}\\\\\n",
      "0 \\text{ otherwise}\n",
      "\\end{cases}$,\n",
      "\n",
      "where:\n",
      "\n",
      "* $I_{(t)}$ is the set of infectious individuals at time $t$\n",
      "* $S_{(t)}$ is the set of susceptible individuals at time $t$\n",
      "* $\\Omega_{S}(i)$ represents risk factors associated with susceptible individual $i$\n",
      "* $\\Omega_{T}(i)$ represents risk factors associated with infectious individual $j$\n",
      "* $\\kappa(i,j)$ is an infection kernel representing risk factors involving both a susceptible individual $i$, and an infectious individual $j$\n",
      "* $\\epsilon(i,t)$ represent external risk factors associated with susceptible individual $i$\n",
      "\n",
      "## Pathogen transmission pathways\n",
      "\n",
      "In an ILM, a susceptible individual's exposure risk is the sum of various risk factors. The specific source of an infection in this framework, however, is irrelevant. In the phylodynamic ILM extension, the source of an infection is of relevance to the both the pathways of pathogen transmission and of pathogen evolution. As such, it is required that seperate exposure probabilities are described for each susceptible-infectious individual combination. Furthermore, as an individual may only be infected from a single source, we must determine which of the competiting sources (if any) are responsible for exposure to the pathogen. To do this we are required to model in continuous time, resulting in the following exposure probability:\n",
      "\n",
      "$P(i, j, t) = \\begin{cases} \\Omega_{S}(i)\\Omega_{T}(j)\\kappa(i,j) \\exp \\left\\{ -\\Omega_{S}(i)\\Omega_{T}(j)\\kappa(i,j)\\delta(t)\\right\\} \\text{ if } i \\in S_{(t)}, j \\in  I_{(t)} \\\\\n",
      "0 \\text{ otherwise}\n",
      "\\end{cases}\n",
      "$,\n",
      "\n",
      "where $\\delta(t)$ is the time from the previous event. Which is to say that an individual's time to exposure is exponentially distributed. An individual's risk of exposure from external factors can also be defined separately as:\n",
      "\n",
      "$P(i, t) = \\begin{cases} \\epsilon(i,t) \\exp \\left \\{ \\epsilon(i,t) \\delta(t) \\right \\} \\text{ if } i \\in S_{(t)} \\\\\n",
      "0 \\text{ otherwise}\n",
      "\\end{cases}\n",
      "$.\n",
      "\n",
      "An individual's transition from an exposed to an infectious state is assumed to be exponentially distributed with probability:\n",
      "\n",
      "$P(i, t) = \\begin{cases} \\rho(i) \\exp \\left \\{ \\rho(i) \\delta(t) \\right \\} \\text{ if } i \\in E_{(t)} \\\\\n",
      "0 \\text{ otherwise}\n",
      "\\end{cases}\n",
      "$.\n",
      "\n",
      "where,\n",
      "\n",
      "* $E_{(t)}$ is the set of exposed individuals at time $t$,\n",
      "* $\\rho(i)$ is the infectivity rate (inverse of the mean latent period) associated with an exposed individual $i$.\n",
      "\n",
      "Similarly, an individual's transition from an infectious state to a removed state is assumed to be exponentially distributed with probability:\n",
      "\n",
      "$P(i, t) = \\begin{cases} \\gamma(i) \\exp \\left \\{ \\gamma(i) \\delta(t) \\right \\} \\text{ if } i \\in I_{(t)} \\\\\n",
      "0 \\text{ otherwise}\n",
      "\\end{cases}\n",
      "$.\n",
      "\n",
      "where,\n",
      "\n",
      "* $\\gamma(i)$ is the removal rate (inverse of the mean infectious period) associated with an infectious individual $i$.\n",
      "\n",
      "## Pathogen phylogenetics\n",
      "\n",
      "Pathogen.jl models mutations in the genome of a pathogen as it is transmitted through a population. In the simple example, this may be done purely using a transition-transversion rate matrix, \n",
      "\n",
      "$\n",
      "\\mathbf{Q} = \\begin{Bmatrix} \n",
      "-\\mu_{A} & \\mu_{AT} & \\mu_{AC} & \\mu_{AG} \\\\\n",
      "\\mu_{TA} & -\\mu_{T} & \\mu_{TC} & \\mu_{TG} \\\\\n",
      "\\mu_{CA} & \\mu_{CT} & -\\mu_{C} & \\mu_{CG} \\\\\n",
      "\\mu_{GA} & \\mu_{GT} & \\mu_{GC} & -\\mu_{G}\n",
      "\\end{Bmatrix},\n",
      "$\n",
      "\n",
      "where each $\\mu$ is the rate of transition or tranversion between two specified nucleotides. These rates must be defined in the same time units as the rates associated with the pathogen transmission process. There are several transition-transversion models that may be utilized of various complexity. Additionally, site heterogeneity could be incorporated into the model as well as other types of mutations. \n",
      "\n",
      "# Simulation framework\n",
      "\n",
      "As individuals in a simulated population are in interaction with one another, the occurence of one event impacts the occurence of other events. Every change in disease state membership (i.e. membership in sets $S$, $E$, $I$, and $R$) impacts event probabilities in some manner. \n",
      "\n",
      "An exposure event to an individual $i$, changes their (previously zero) probability of becoming infectious and also zeros their probability of becoming exposed from any of the potential sources. Once an individual has been exposed there are also probabilities associated with the mutation of the pathogen that has been transmitted to them, which remains non-zero until individual $i$ is removed from the population.\n",
      "\n",
      "An infection event to an individual $i$, changes their (previously zero) probability of being removed and also zeros their probability of becoming infectious. While individual $i$ is infectious, every remaining susceptible individual in the population are at some risk of exposure to the pathogen that individual $i$ hosts.\n",
      "\n",
      "Under the assumption of an SEIR model, a removal event to individual $i$ zeros their probability of removal, and also zeros the probability that they may expose any susceptible individual.\n",
      "\n",
      "A mutation event may only change that nucleotide's mutation rate (i.e. under a neutral theory of pathogen evolution), it is also possible that a pathogen sequence could be associated with disease transmissability, latency, or recovery, creating potential for selection pressures.\n",
      "\n",
      "The most straightforward and computationally efficient method of generating these interdependent events involves the minimum of independent exponential random variables. As the occurence of each event is described by an exponential distribution, we can generate the minimum of these distributions through summing their rate parameters. That is, the time of the next event in our system can be assumed to be exponentially distributed, with rate\n",
      "\n",
      "$ \\lambda = \\sum_{i \\in S_{(t)}} \\sum_{j \\in  I_{(t)}} \\left[ \\Omega_{S}(i)\\Omega_{T}(j)\\kappa(i,j) \\right] + \\sum_{i \\in S_{(t)}} \\left[ \\epsilon(i,t) \\right] + \\sum_{i \\in E_{(t)}} \\left[ \\rho(i) \\right] + \\sum_{i \\in I_{(t)}} \\left[ \\gamma(i) \\right] $\n",
      "\n",
      "The event that occurs at this time can be found through sampling from a discrete distribution. The proportion of $\\lambda$ that can be attributed to any event, is its probability in being selected. After an event time has been generated, and a specific event type selected, the necessary rate updates can be made. In Pathogen.jl this information is conveinently organized and modified in a rate array where each column represents an individual (or an \"external source\"), and each row an event type. Note: \n",
      "\n",
      "# Inference framework\n",
      "\n",
      "## Data augmentation\n",
      "\n",
      "# Future work\n",
      "\n",
      "# References"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}